aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,338 @@
+    }
+
+    struct DepositEvent has drop, store {
+        object_id: address,
+    }
+
+    struct WithdrawEvent has drop, store {
+        object_id: address,
+    }
+
+    public entry fun init_store(account: &signer) {
+        move_to(
+            account,
+            ObjectStore {
+                inner: table::new(),
+                deposits: account::new_event_handle(account),
+                withdraws: account::new_event_handle(account),
+            },
+        )
+    }
+
+    public fun withdraw(account: &signer, addr: address): OwnerAbility acquires ObjectStore {
+        let object_store = borrow_global_mut<ObjectStore>(signer::address_of(account));
+        event::emit_event(&mut object_store.withdraws, WithdrawEvent { object_id: addr });
+        table::remove(&mut object_store.inner, addr)
+    }
+
+    public fun deposit(account: &signer, object: OwnerAbility) acquires ObjectStore {
+        let object_store = borrow_global_mut<ObjectStore>(signer::address_of(account));
+        let object_addr = owner_ability_address(&object);
+        event::emit_event(&mut object_store.deposits, DepositEvent { object_id: object_addr });
+        table::add(&mut object_store.inner, object_addr, object);
+    }
+
+    public fun deposit_typed<T: key>(
+        account: &signer,
+        object: TypedOwnerAbility<T>,
+    ) acquires ObjectStore {
+        deposit(account, to_owner_ability(object))
+    }
+
+    public fun contains(account: &signer, addr: address): bool acquires ObjectStore {
+        let object_store = borrow_global<ObjectStore>(signer::address_of(account));
+        table::contains(&object_store.inner, addr)
+    }
+
+    /// Object addresses are equal to the sha3_256([creator addres | seed | 0xFE]).
+    public fun create_object_id(source: &address, seed: vector<u8>): address {
+        let bytes = bcs::to_bytes(source);
+        vector::append(&mut bytes, seed);
+        vector::push_back(&mut bytes, OBJECT_ADDRESS_SCHEME);
+        from_bcs::to_address(hash::sha3_256(bytes))
+    }
+
+    public fun create_object(creator: &signer, seed: vector<u8>): CreatorAbility {
+        // create object address needs to be distinct from create_resource_address
+        let address = create_object_id(&signer::address_of(creator), seed);
+        assert!(!exists<Object>(address), error::already_exists(EOBJECT_EXISTS));
+        let object_signer = create_signer_internal(address);
+
+        move_to(
+            &object_signer,
+            Object {
+                guid_creation_num: 0,
+            },
+        );
+        CreatorAbility { self: address }
+    }
+
+    public fun create_guid(object: &signer): guid::GUID acquires Object {
+        let addr = signer::address_of(object);
+        let object_data = borrow_global_mut<Object>(addr);
+        guid::create(addr, &mut object_data.guid_creation_num)
+    }
+
+    public fun generate_delete_ability(ability: &CreatorAbility): DeleteAbility {
```

Comment:
So creators can create multiple copies of SignerAbility, OwnerAbility, and DeleteAbility? Why do we want to support this vs a single copy of each? Having multiple ownership of the same object sounds complicated and can lead to more edge cases.